File: native-linux/src/data/network/karakeep_client.cpp
#include "karakeep_client.h"

#include "data/network/http_soup.h"
#include "data/prefs/prefs.h"

#include <algorithm>
#include <map>
#include <stdexcept>

#include <nlohmann/json.hpp>

namespace cd {

KarakeepClient::KarakeepClient(SecretStore& secrets, SoupSession* session)
    : secrets_(secrets)
    , session_(session)
{
    if (!session_) throw std::invalid_argument("KarakeepClient: null SoupSession");
}

std::vector<KarakeepFolder> KarakeepClient::list_folders() const
{
    auto json = nlohmann::json::parse(request("GET", "lists"));
    std::vector<KarakeepFolder> folders;
    for (auto const& item : json.value("lists", nlohmann::json::array())) {
        if (!item.is_object() || item.value("type", "manual") != "manual") continue;
        std::string role = item.value("userRole", "owner");
        if (role != "owner" && role != "editor") continue;
        KarakeepFolder folder;
        folder.id = item.value("id", "");
        folder.name = item.value("name", "");
        if (item.contains("parentId") && item["parentId"].is_string())
            folder.parent_id = item["parentId"].get<std::string>();
        if (!folder.id.empty() && !folder.name.empty()) folders.push_back(std::move(folder));
    }
    std::sort(folders.begin(), folders.end(), [](auto const& left, auto const& right) {
        return left.name < right.name;
    });
    return folders;
}

void KarakeepClient::save_urls(
    std::vector<std::string> const& urls, std::optional<std::string> const& folder_id) const
{
    for (auto const& url : urls) {
        nlohmann::json payload = {
            {"type", "link"},
            {"url", url},
            {"source", "api"},
        };
        auto response = nlohmann::json::parse(request("POST", "bookmarks", payload.dump()));         std::string bookmark_id = response.value("id", "");
        if (bookmark_id.empty()) throw std::runtime_error("Karakeep returned no bookmark ID");
        if (folder_id.has_value()) {                                                                     (void)request(                                                                                   "PUT", "lists/" + *folder_id + "/bookmarks/" + bookmark_id, std::string{}, true);                                                                                                 }
    }
}

std::string KarakeepClient::api_base() const
{
    auto host = secrets_.get(CredentialKey::KARAKEEP_HOST);
    if (!host.has_value() || host->empty()) throw std::runtime_error("Karakeep server URL is missing");
    while (!host->empty() && host->back() == '/') host->pop_back();
    return *host + "/api/v1/";
}

std::string KarakeepClient::request(char const* method, std::string const& path,                 std::optional<std::string> const& body, bool allow_empty) const
{
    auto token = secrets_.get(CredentialKey::KARAKEEP_TOKEN);
    if (!token.has_value() || token->empty()) throw std::runtime_error("Karakeep API key is missing");
    std::map<std::string, std::string> headers = {                                                   {"Accept", "application/json"},
        {"Authorization", "Bearer " + *token},                                                   };
    auto [status, response] = soup_sync_request(
        session_, method, api_base() + path, body, body.has_value() ? "application/json" : nullptr, headers);
    if (status < 200 || status >= 300)
        throw std::runtime_error("Karakeep returned HTTP " + std::to_string(status));
    if (response.empty() && !allow_empty) throw std::runtime_error("Karakeep returned an empty response");
    return response;
}
} // namespace cd

File: native-android/app/src/main/kotlin/com/crossdashboard/app/data/network/KarakeepClient.kt
package com.crossdashboard.app.data.network

import com.crossdashboard.app.data.prefs.CredentialKey
import com.crossdashboard.app.data.prefs.SecureStore
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.IOException
import javax.inject.Inject
import javax.inject.Singleton

@Serializable
data class KarakeepFolder(
    val id: String,
    val name: String,
    val parentId: String? = null,
    val type: String = "manual",
    val userRole: String = "owner",
)

@Serializable
private data class KarakeepFolderResponse(val lists: List<KarakeepFolder> = emptyList())

@Serializable
private data class KarakeepBookmarkResponse(val id: String)

@Serializable
private data class CreateKarakeepBookmarkRequest(
    val type: String = "link",
    val url: String,
    val source: String = "api",
)

@Singleton
class KarakeepClient @Inject constructor(
    private val secureStore: SecureStore,
    private val httpClient: OkHttpClient,
) {
    private val json = Json { ignoreUnknownKeys = true }

    suspend fun listFolders(): List<KarakeepFolder> = withContext(Dispatchers.IO) {
        val response = execute(Request.Builder().url("${apiBase()}/lists").get().build())
        json.decodeFromString<KarakeepFolderResponse>(response).lists
            .filter { it.type == "manual" && it.userRole in setOf("owner", "editor") }
            .sortedBy { it.name.lowercase() }
    }

    suspend fun saveUrls(urls: List<String>, folderId: String?) = withContext(Dispatchers.IO) {
        urls.distinct().forEach { url ->
            val payload = json.encodeToString(CreateKarakeepBookmarkRequest(url = url))
            val response = execute(
                Request.Builder()
                    .url("${apiBase()}/bookmarks")
                    .post(payload.toRequestBody(JSON_MEDIA_TYPE))
                    .build(),
            )
            val bookmarkId = json.decodeFromString<KarakeepBookmarkResponse>(response).id
            if (folderId != null) {
                execute(
                    Request.Builder()
                        .url("${apiBase()}/lists/$folderId/bookmarks/$bookmarkId")
                        .put(ByteArray(0).toRequestBody(null))
                        .build(),
                    allowEmpty = true,
                )
            }
        }
    }

    private fun apiBase(): String {
        val host = secureStore.get(CredentialKey.KARAKEEP_HOST)
            ?.trim()
            ?.trimEnd('/')
            ?.takeIf { it.isNotEmpty() }
            ?: throw IOException("Karakeep server URL is missing")
        return "$host/api/v1"
    }

    private fun execute(request: Request, allowEmpty: Boolean = false): String {
        val token = secureStore.get(CredentialKey.KARAKEEP_TOKEN)
            ?.takeIf { it.isNotBlank() }
            ?: throw IOException("Karakeep API key is missing")
        val authenticated = request.newBuilder()
            .header("Accept", "application/json")
            .header("Authorization", "Bearer $token")
            .build()
        return httpClient.newCall(authenticated).execute().use { response ->
            if (!response.isSuccessful) {
                throw IOException("Karakeep returned HTTP ${response.code}")
            }
            response.body?.string().orEmpty().also {
                if (!allowEmpty && it.isEmpty()) throw IOException("Karakeep returned an empty response")
            }
        }
    }

    private companion object {
        val JSON_MEDIA_TYPE = "application/json".toMediaType()
    }
}

